p2p/ipld/write.go

```diff
@@ -42,12 +52,113 @@ func PutBlock(ctx context.Context, adder ipld.NodeAdder, block *types.Block) err
 	if err != nil {
 		return fmt.Errorf("failure to recompute the extended data square: %w", err)
 	}
+	// get row and col roots to be provided
+	// this also triggers adding data to DAG
+	prov := newProvider(ctx, croute, int32(squareSize*4), logger.With("height", block.Height))
+	for _, root := range eds.RowRoots() {
+		prov.Provide(plugin.MustCidFromNamespacedSha256(root))
+	}
+	for _, root := range eds.ColumnRoots() {
+		prov.Provide(plugin.MustCidFromNamespacedSha256(root))
+	}
+	// commit the batch to ipfs
+	err = batchAdder.Commit()
+	if err != nil {
+		return err
+	}
+	// wait until we provided all the roots if requested
+	<-prov.Done()
+	return prov.Err()
+}
 
-	// thanks to the batchAdder.Visit func we added to the nmt wrapper,
-	// generating the roots will start adding the data to IPFS
-	eds.RowRoots()
-	eds.ColumnRoots()
+var provideWorkers = 32
 
-	// commit the batch to ipfs
-	return batchAdder.Commit()
+type provider struct {
+	ctx  context.Context
+	done chan struct{}
+
+	err   error
+	errLk sync.RWMutex
+
+	jobs  chan cid.Cid
+	total int32
+
+	croute routing.ContentRouting
+	log    log.Logger
+	took   time.Time
+}
+
+func newProvider(ctx context.Context, croute routing.ContentRouting, toProvide int32, logger log.Logger) *provider {
+	p := &provider{
+		ctx:    ctx,
+		done:   make(chan struct{}),
+		jobs:   make(chan cid.Cid, provideWorkers),
+		total:  toProvide,
+		croute: croute,
+		log:    logger,
+	}
+	for range make([]bool, provideWorkers) {
+		go p.worker()
+	}
+	logger.Info("Started Providing to DHT")
+	p.took = time.Now()
+	return p
+}
+
+func (p *provider) Provide(id cid.Cid) {
+	select {
+	case p.jobs <- id:
+	case <-p.ctx.Done():
+	}
+}
+
+func (p *provider) Done() <-chan struct{} {
+	return p.done
+}
+
+func (p *provider) Err() error {
+	p.errLk.RLock()
+	defer p.errLk.RUnlock()
+	if p.err != nil {
+		return p.err
+	}
+	return p.ctx.Err()
+}
+
+func (p *provider) worker() {
+	for {
+		select {
+		case id := <-p.jobs:
+			err := p.croute.Provide(p.ctx, id, true)
+			if err != nil && err != kbucket.ErrLookupFailure { // Check for error to decrease test log spamming
```

Comment:
That error was introduced [7 years ago](https://github.com/libp2p/go-libp2p-kbucket/commit/2f8828de36de86c0840d8d4901ccfacee4b808f6) before major DHT refactor/rework and the reason for having it exported is kinda out of date IMO. Right now, this error is solely a symptom of a node not having any connected DHT nodes and in such a case there would other symptoms that things go wrong, so we can omit to rely on this error. Otherwise, tests are spammed a lot with it for cases where we do not care about providing. The proper solution is to extract providing at all and as I mentioned in the root PR comment, the providing code is not final.

Reply:
alrighty, sounds good. Mind documenting that?